Index: sites/all/modules/_mobile/commerce_services/commerce_services.module
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- sites/all/modules/_mobile/commerce_services/commerce_services.module	(revision )
+++ sites/all/modules/_mobile/commerce_services/commerce_services.module	(revision )
@@ -1942,7 +1942,7 @@
   list(, , $bundle) = entity_extract_ids($entity_type, $entity);
 
   // Look for each type of Commerce reference field on the given entity.
-  foreach (array('commerce_product_reference', 'commerce_line_item_reference', 'commerce_customer_profile_reference', 'taxonomy_term_reference') as $field_type) {
+  foreach (array('field_collection', 'commerce_product_reference', 'commerce_line_item_reference', 'commerce_customer_profile_reference', 'taxonomy_term_reference') as $field_type) {
     // Look for reference field instances of each type on the entity.
     foreach (field_info_instances($entity_type, $bundle) as $field_name => $instance) {
       $entities_field_name = $field_name . '_entities';
